File: 0697. Degree of an Array/0697.java
// time complexity: O(n)
// space complexity: O(n)
class Solution {
    public int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> count = new HashMap<>();       // number - frequency
        Map<Integer, Integer> first = new HashMap<>();       // number - first index

        int degree = 0;
        int minLen = nums.length;
        for(int i=0; i<nums.length; i++){
            first.putIfAbsent(nums[i], i);                   // record first occurrence
            count.put(nums[i], count.getOrDefault(nums[i], 0) + 1);   // increment count

            // update degree and minimum length accordingly
            int freq = count.get(nums[i]);
            if (freq > degree) {
                degree = freq;
                minLen = i - first.get(nums[i]) + 1;
            }
            else if (freq == degree) {
                minLen = Math.min(minLen, i - first.get(nums[i]) + 1);
            }
        }
        return minLen;
    }
}

File: 0697. Degree of an Array/0697.py
# time complexity: O(n)
# space complexity: O(n)
class Solution(object):
    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # count frequency of each element
        count = Counter(nums)
        degree = max(count.values())

        # record first and last occurrence of each element
        left, right = {}, {}
        for i, x in enumerate(nums):
            left.setdefault(x, i)
            right[x] = i

        # compute minimal length among elements with max frequency
        min_len = len(nums)
        for x, freq in count.items():
            if freq == degree:
                curr_len = right[x] - left[x] + 1  # window size for x
                if curr_len < min_len:
                    min_len = curr_len

        return min_len